tests/test_solver_regression.py

```diff
@@ -0,0 +1,86 @@
+"""Regression tests for solver bugs found during development."""
+
+import numpy as np
+import pytest
+from scipy.sparse import csr_matrix
+
+from MCEq.solvers import solv_numpy
+from MCEq import config
+
+
+def test_solv_numpy_does_not_modify_input_phi():
+    """
+    Regression test: ensure solv_numpy doesn't modify the input phi array.
+    
+    This was a bug where the NumPy solver modified the input array in place,
+    causing subsequent solver calls to start with wrong initial conditions.
+    """
+    # Create test data
+    nsteps = 10
+    size = 5
+    dX = np.full(nsteps, 0.1)
+    rho_inv = np.ones(nsteps)
+    grid_idcs = list(range(nsteps))
+    
+    lam_int = 0.3
+    lam_dec = 0.1
+    int_m = csr_matrix(-lam_int * np.eye(size))
+    dec_m = csr_matrix(-lam_dec * np.eye(size))
+    
+    phi_original = np.ones(size)
+    phi_copy = phi_original.copy()
+    
+    # Run solver
+    solution, _ = solv_numpy(nsteps, dX, rho_inv, int_m, dec_m, phi_original, grid_idcs)
+    
+    # The input array should not be modified
+    np.testing.assert_array_equal(
+        phi_original, phi_copy,
+        err_msg="solv_numpy modified the input phi array - this breaks subsequent solver calls"
+    )
+    
+    # The solution should be different from the input
+    assert not np.array_equal(solution, phi_copy), "Solver should produce a different result"
+
+
+@pytest.mark.skipif(not config.has_cuda, reason="CUDA not available")
+def test_cuda_numpy_solver_consistency():
+    """
+    Regression test: ensure CUDA and NumPy solvers produce consistent results.
+    
+    This validates that both solvers implement the same algorithm correctly.
+    
+    Note: Requires CuPy >= 12.0.0 for modern sparse matrix interface compatibility.
+    """
+    from MCEq.solvers import solv_CUDA_sparse, CUDASparseContext
+    from MCEq import config
+    
+    # Create test data
+    nsteps = 10
+    size = 5
+    dX = np.full(nsteps, 0.1)
+    rho_inv = np.ones(nsteps)
+    grid_idcs = list(range(nsteps))
+    
+    lam_int = 0.3
+    lam_dec = 0.1
+    int_m = csr_matrix(-lam_int * np.eye(size))
+    dec_m = csr_matrix(-lam_dec * np.eye(size))
+    
+    phi_numpy = np.ones(size)
+    phi_cuda = np.ones(size)
+    
+    # Run NumPy solver
+    solution_numpy, _ = solv_numpy(nsteps, dX, rho_inv, int_m, dec_m, phi_numpy, grid_idcs)
+    
+    # Run CUDA solver  
+    ctx = CUDASparseContext(int_m, dec_m, device_id=config.cuda_gpu_id)
+    solution_cuda, _ = solv_CUDA_sparse(nsteps, dX, rho_inv, ctx, phi_cuda, grid_idcs)
+    
+    # Results should match within floating-point precision
+    # (CUDA uses float32, NumPy uses float64, so we expect small differences)
+    np.testing.assert_allclose(
+        solution_cuda, solution_numpy,
+        rtol=1e-5, atol=1e-8,
+        err_msg="CUDA and NumPy solvers produce different results"
+    )
```

Comment:
see above comments. Do we need the advanced stats from `assert_allclose`? `assert np.allclose(...)` still prints the complete arrays and is easier to read (at least for me). What do you think? 😊

Reply:
Could you implement these non-essential changes and push to this PR? I'm fine with those proposals